### PR TITLE
AWSReader methods return std error rather than specific error type

### LINKED ISSUE #12: Public interfaces should return std error rather than custom type
Returning [custom error type](https://github.com/cycloidio/raws/blob/master/connector.go#L31) rather [standard error](https://golang.org/pkg/builtin/#error) in public interfaces is inconvenient.

First reason is that it forces to other implementations to use the same error type although it satisfies the [standard error interface](https://golang.org/pkg/builtin/#error) and the return errors shouldn't be a strong requirement for public interfaces, rather than being a standard error; on the other hand it's idiomatic to return standard errors.

Second reason is that when the returned error is `nil`, it cannot be coerced to a [standard error](https://golang.org/pkg/builtin/#error) without losing the the ability to be compare with `nil`.

For example (simple silly one in terms of logic), it cannot be done something like this

```go
func myInstances() ([]*ec2.DescribeInstancesOutput, error) {
  	// Do whatever initializations or other custom operations here
  	// in order to get the AWS Reader implementation instance, which
  	// below is set to the variable `c`
 
	return c.GetInstances(nil)
}

// In other place where the call to such function is done
var instances, err = myInstances()

// although the `c.GetInstances` returned `nil`, `err` isn't nil, because it has been coerced
// `error` producing a interface which value isn't nil, it's a pointer to the original error type 
// which value is `nil`, making it incomparable to `nil`.
if err != nil {
	os.Exit(1)
}

fmt.Println(instances)
```

The big problem is that it's very easy to get trap with the coercion issue and struggling to understand why you're getting an error when actually, you aren't getting any, and makes the library harder to use.
A correct usage of the library for the previous example is:

```go
func myInstances() ([]*ec2.DescribeInstancesOutput, error) {
	// Do whatever initializations or other custom operations here
  	// in order to get the AWS Reader implementation instance, which
  	// below is set to the variable `c`
 
 	var instances, err = c.GetInstances(nil)
	if err != nil {
		return nil, err
	}

	return instances, nil
 }

// In other place where the call to such function is done
var instances, err = myInstances()
if err != nil {
  os.Exit(1)
}

fmt.Println(instances)
```

This playground example show the actual case https://play.golang.org/p/vHN0Oxjylmd

If it looks good to you for making the chance, I could send a PR to make the changes, although those changes will be breaking changes, some consumers may break if they have been using the specific returned error type.

### PR DESCRIPTION
**Pull request by ifraixedes**
In order to drive the refactoring by test failures and making the appropriated changes to fix the test failures, I needed to base this branch on the changes done in #24 hence, first review #24 before this one and once that's merged, I'll rebase this one from master.

About the changes of this branch, the subject of this PR with its associated issue #12 and the commit messages should be enough to know about it.

If you prefer to squash the commits of the refactoring of the methods (done, method by method), let me know and I'll do it just before merging for avoiding conflicts on rebases from the last master.

closes #12

### CODE REVIEW

File: error.go [hunk 1]
```diff
@@ -45,27 +67,90 @@ func (e Errs) Error() string {
...
+		if e.Region() == region {
+			return err
+		}
 
-// Returns the region of the error
-func (e *callErr) Region() string {
-	return e.region
-}
+		return nil
+	case *Error:
```
Comment (xescugc) [inline error.go:85]: IDK if we should handle this case IMO an `error` should never be a pointer.

File: error.go [hunk 2]
```diff
@@ -45,27 +67,90 @@ func (e Errs) Error() string {
...
+		}
+
+		return nil
+	case Errors:
+		var errs Errors
+
+		for _, ei := range e {
+			if ei.Region() == region {
+				errs = append(errs, ei)
+			}
```
Comment (xescugc) [inline error.go:101]: I would use recursion for this, just call `ErrorIn(region, e)` and append if not nil. Would make the code more reusable, this way you only have the logic on one `case`, the `Error`

File: error.go [hunk 3]
```diff
@@ -45,27 +67,90 @@ func (e Errs) Error() string {
...
+// in the region, or a value of the type Errors when multiple errors are found.
+func ErrorFrom(service string, err error) error {
+	switch e := err.(type) {
+	case Error:
+		if e.Service() == service {
+			return err
+		}
+
+		return nil
+	case *Error:
```
Comment (xescugc) [inline error.go:129]: Same I commented befor.

File: error.go [hunk 4]
```diff
@@ -45,27 +67,90 @@ func (e Errs) Error() string {
...
+
+		return nil
+	case Errors:
+		var errs Errors
+
+		for _, ei := range e {
+			if ei.Service() == service {
+				errs = append(errs, ei)
+			}
+		}
```
Comment (xescugc) [inline error.go:146]: Same comment about recursion.

**Review by xescugc — CHANGES_REQUESTED** [review]
I did miss the #12 so I'll give my opinion here. This [post](https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully) explaing exatly the issue we are having, and the actual solution it's good. I would suggest to use the github.com/pkg/errors to wrap the errors as it gives a good api to extend them and add more information.

This would involve to change everything so not now, but just mentioning it.

File: error.go (hunk 1 above)
Comment (ifraixedes) [reply error.go:85]: I prefer to keep it

File: error.go (hunk 3 above)
Comment (ifraixedes) [reply error.go:129]: I prefer to keep it

File: error.go (hunk 4 above)
Comment (ifraixedes) [reply error.go:146]: The logic is tiny, it's basically a condition and doing a recursion doesn't convince me

```
      for _, ei := range e 
		if ef := ErrorFrom(service, ei); ef != nil {
         	errs = append(errs, ef.(Error))
        }
      }
```

There is a conditional anyway and a type assertion; considering the replicated logic, I see that's more readable the loop version.

Without considering that, historically, Go behaves better in loops than recursion calls and that the each recursive call has to execute the `switch type` again, however these 2 things are not an argument without providing benchmarks.

File: error.go (hunk 2 above)
Comment (ifraixedes) [reply error.go:101]: Same than mentioned in https://github.com/cycloidio/raws/pull/25#discussion_r166856366

**Review by xescugc — APPROVED** [review]
About the `*Error` for me it's a smelly of bad design that an error is a pointer, that's why I mention to remove it, but ok.

About the recursion, as it's small now it seems useless to add this, but in the future if you have to change the code you'll have to modify the same code on 2 different places.

You could abstract the logic on a private method, but for now if you prefer keep it this way.

**Comment by ifraixedes** [comment]
I don't see any clear future that those 2 functions will have a more complicated logic.

> About the *Error for me it's a smelly of bad design that an error is a pointer, that's why I mention to remove it, but ok.

I don't see it as if you want to use the type `Error` on whatever place and you want to use it as a pointer for being able to return `nil`, the function will work anyway.

**Comment by xescugc** [comment]
> I don't see it as if you want to use the type Error on whatever place and you want to use it as a pointer for being able to return nil, the function will work anyway.

All the internal functions (IRC) use `error` interface as signature. So you can send `nil` already, the use case that you define, IMHO, seems really wrong code that's why it should not be handled.

**Comment by ifraixedes** [comment]
For example, in the case that a consumer of the library define its own error for some need that it has and embed the `Error` as in the following example 

```
type MyErr {
    inner *Error
    code int
}

func (m MyErr) Error() string {
 // Whatever here
}
```

and it wants to use the function to do whatsoever, there is no need to check for if `inner` is nil before calling `ErrorFom` because you couldn't do `ErrorFrom(*inner)` without checking first is `inner` isn't `nil` otherwise it would be a panic, so I see that it's handy to be able to do `ErrorFrom(inner)` as pointer is accepted.

I don't see that's really wrong as you mentioned.

**Comment by xescugc** [comment]
Well we disagree with it :), because a use could pass to `ErrorFrom` any impmementation of the `error` interface so :), but well  it's ok.

**Comment by ifraixedes** [comment]
yes, but other implementations of error aren't in this package.
Disagreement isn't a problem, but I don't think that there is an strong argument to say that's really wrong.

**Pull request merged**